### PR TITLE
Add multivariate output fitting tool

### PR DESCRIPTION
**Pull request by perrygreenfield**
Since many transforms between coordinate systems involve two input coordinates and two output coordinates, there is a need to adapt astropy modeling to be able to handle more than one output value to be fit. This provides such a tool that will accept an astropy model and astropy fitter instance along with the input values and output values to fit.

### CODE REVIEW

File: tweakwcs/multivariate.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    for input in inputs:
+        newinput = input.copy()
+        newinput.shape = (1, input.shape[0])
+        expanded_inputs.append(np.ravel(expander * newinput))
+    newoutputs = np.concatenate(outputs)
+    # Handle the same with weights, if present
+    if 'weights' in kwargs:
+        newweights = np.ravel(expander * weights)
+    else:
+        newweights = None
```
Comment (mcara) [inline tweakwcs/multivariate.py:95]: `weights` is undefined. I suggest:

```suggestion
    newweights = kwargs.get('weights', None)
    if weights:
        newweights = np.ravel(expander * newweights)
```
In addition, add a unit test to cover this case.

File: tweakwcs/multivariate.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+    # Determine number of inputs and ouputs of model
+    n_inputs = model.n_inputs
+    n_outputs = model.n_outputs
+    if n_outputs < 2:
+        raise ValueError(
+            "Pointless to use this fitting function for less than 2 output values")
+    # Assumes all inputs and ouputs are 1D arrays and aleady broadcasted
+    # Generate input repetition (based on number of outputs)
+    expanded_inputs = []
+    expander = np.ones((n_outputs, 1))
```
Comment (mcara) [inline tweakwcs/multivariate.py:85]: I do not understand the need for `expander` which is later used in conjunction with `ravel`. Can it be replaced with some form of reshape or new axis? [Just a question]

File: tweakwcs/multivariate.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+        size = inp_array.shape[0]
+        return np.arange(size)
+
+    def inverse(self):
+        raise NotImplementedError('Index model has no inverse by design')
+
+
+def _multi_output_fit(fitter, model, inputs, outputs, **kwargs):
+    """
+    Function to enable fits to models with more than one output variable.
```
Comment (mcara) [inline tweakwcs/multivariate.py:52]: Could you, please, add some more details to the description like what does this model actually do with multiple outputs? How will the _actual_ fitting be performed? _"Function to enable fits..."_ seems too vague.

File: CHANGELOG.rst [hunk 4]
```diff
@@ -7,6 +7,9 @@ Release Notes
 .. 0.7.3 (unreleased)
    ==================
 
+- Added (temporarily) a module to allow multivariate output fits.
+  (multivariate.py) [#137]
```
Comment (mcara) [inline CHANGELOG.rst:11]: No need to mention something temporary in the change log. Let's keep this under the radar.

File: tweakwcs/multivariate.py [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+Notes
+-----
+
+This presumes that the multiple outputs are given equal weight in
+computing the least squares value. A future version may allow 
+specifying different weights when the quantities involved are
+not directly comparable. The motivating case for developing this
+was to compute the proper transformation of 2D coordinates, and as
+such, the least square computation is equivalent to computing the
+least squares radial differences in position. This could be generalized
```
Comment (mcara) [inline tweakwcs/multivariate.py:14]: **radial differences** sounds confusing: is it a list of differences like `x1 - x1_0`, `x2 - x2_0`, ...,  `y1 - y1_0`, `y2 - y2_0`, ... **OR** is it like `r1 - r1_0`, `r2 - r2_0`, ... ?

**Comment by stsci-hack** [comment]
The main driver for this PR was the inability of the current astropy.modeling support for effectively replicating what 'tweakwcs' does; specifically, performing an 'rscale' fit between an input set of coordinates and a reference catalog of coordinates.  The astropy.modeling notebook used for JWSTUserTraining introduced the use of 2D Polynomials for fitting, but did not actually use them with a fitter because they did not work like the 1D case used earlier in the notebook.  The changes in this PR allow the type of fitting of 2D coordinates from the notebook to actually be fit using a fitter. 

We need this functionality in astropy.modeling so that we can expand on the models we can fit beyond what 'tweakwcs' can do as 'tweakwcs' can only fit for linear global transformations between catalogs (coordinate frames). The ultimate goal is to be able to fit proper motions for each source in addition to the global transformations all at the same time which astropy.modeling can support once the problems with fitting addressed by this PR are implemented.

**Comment by mcara** [comment]
> ... use of 2D Polynomials for fitting ...

I believe this particular fitting _is supported_: a 2D polynomial maps 2D domain to a 1D output. I believe the issue is when you want to fit **vector-valued** functions to, e.g., 2D data points.

**Comment by mcara** [comment]
> ... use of 2D Polynomials for fitting ...

I believe this particular fitting _is supported_: a 2D polynomial maps 2D domain to a 1D output. I believe the issue is when you want to fit **vector-valued** functions to, e.g., 2D data points.

For example, @perrygreenfield code in `gwcs.WCS.to_fits_sip()` already fits 2D Polynomials (SIP) to data points.

**Comment by mcara** [comment]
If you need to fit 2D-vector-models to 2D points, I have a different suggestion that works with existing `astropy` modeling infrastructure and is significantly simpler than proposed approach. This is an example based on your unit test. The main code is in second block `In [81]`:

```python
In [80]: import numpy as np
    ...: from astropy.modeling.models import Shift, Rotation2D
    ...: from astropy.modeling.fitting import LevMarLSQFitter, _fitter_to_model_params

In [81]: inputs = [np.array([10., 10., 20., 20.]), np.array([10., 20., 20., 10.])]
    ...: outputs = [np.array([8.06101731, 0.98994949, 8.06101731, 15.13208512]),
    ...:            np.array([12.16223664, 19.23330445, 26.30437226, 19.23330445])]
    ...: r = Rotation2D()
    ...: r.fittable = True
    ...: model = (Shift() & Shift()) | r
    ...:

In [82]: class LMF2x2(LevMarLSQFitter):
    ...:     def objective_function(self, fps, *args):
    ...:         model, weights, inputs, meas = args
    ...:         _fitter_to_model_params(model, fps)
    ...:         return np.ravel(np.subtract(model(*inputs), meas))
    ...:

In [83]: f = LMF2x2()

In [84]: f(model, inputs, outputs)
Out[84]: <CompoundModel(offset_0=4.3, offset_1=-7.1, angle_2=45.)>
```

**Comment by perrygreenfield** [comment]
I must admit this is a much easier way to do it. So let's dump this PR. Warren's eventual use case is more involved, but that is not relevant to this approach. I missed the fact that the objective function can be used to map multiple outputs into a single array.

**Pull request closed without merging**